clang/lib/Frontend/InitPreprocessor.cpp

```diff
@@ -453,8 +453,12 @@ static void InitializeStandardPredefinedMacros(const TargetInfo &TI,
     if (LangOpts.CPlusPlus26)
       // FIXME: Use correct value for C++26.
       Builder.defineMacro("__cplusplus", "202400L");
-    else if (LangOpts.CPlusPlus23)
+    else if (LangOpts.CPlusPlus23) {
       Builder.defineMacro("__cplusplus", "202302L");
+      // [C++23] 15.11p2 [cpp.predefined]
+      Builder.defineMacro("__STDCPP_FLOAT16_T__", "1");
```

Comment:
I'm generally out of my depth in Clang code for the most part, but my recollection is that `_Float16` is currently only supported on "some" architectures (with a cursory check suggesting that "some" is surprisingly hard to make precise), and I'm not seeing anything in this code that makes `_Float16` unconditionally available in C++23 mode.

Reply:
Right there is `TargetInfo::hasFloat16Type` / `hasFullBFloat16Type`.
We probably want to check that both here, and when creating such type during parsing